working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+
+```dart
+class Complex {
+  final double real, imaginary;
+  Complex({this.real, this.imaginary});
+}
+class ColorComplex extends Complex {
+  final Color color;
+  ColorComplex({double real, double imaginary, this.color}) 
+      : super(real: real, imaginary: imaginary);  // "real"/"imaginary" written three times.
+}
+```
+
+It would be nice to have a simpler way to handle simple cases, without making anything harder for complex cases.
+
+### Default Constructors
+
+Dart adds a *default constructor* to any class which doesn't declare any constructors. The default constructor is an unnamed generative constructor which is equivalent to `ClassName() : super();`.
+
+If that constructor would not be valid for the class, for example if the class declares a final instance variable with no initializer, or if the superclass does not have an unnamed generative constructor with no required parameters, then it's a compile-time error.
+
+In most situations, there is no difference between declaring the constructor above explicitly and getting the default constructor. The only distinction is that you can extract a mixin from a class if it has `Object` as superclass and it declares no non-factory constructor.
+
+### Mixin Application Forwarding Constructors
+
+A *mixin application* receives corresponding forwarding constructors for all accessible generative constructors of the superclass. The forwarding constructor has the same basename (identifer after the class name for named constructors, empty for the unnamed constructor), the same parameter signature as the superclass constructor, including default values for optional parameters, and an initializer list with a `super`-invocation to the corresponding superclass constructor which forwards each parameter as an argument to the corresponding superclass constructor parameter. The forwarding constructor is `const` if the superclass constructor is `const` *and* the mixin declares no instance variables.
+
+Since a mixin application class cannot contain static members, there is no risk of the introducing a conflict between a constructor name and a static member. 
+
+### Goal
+
+We combine the forwarding constructor feature with default constructors and automatic initialization of declared instance variables. This *unifies* default constructors and mixin application forwarding constructors *and* adds more power to the default constructor for simple classes.
+
+We expect this to happen after non-nullable types have been added to the language, so the proposal is written in terms of that, with a few guidelines for how legacy libraries work where nullability matters.
+
+## Proposal
+
+Let *C* be a class declaration with name `C` and superclass *S* with name `S` and which *declares* no constructor.
```

Comment:
Now that the default constructor is manifestly useful, I think we should anticipate that users will want to have access to it even in classes that define other constructors. In fact, a useful pattern would probably be to have other constructors that just forward to the now-very-useful default one. It practically replaces field initialization lists, pretty much like primary constructors in Scala and Kotlin.

At the same time, users will sometimes want to hide that default constructor, so if declaring a constructor does not hide it, we'd need some other mechanism to do so.

5-minute bikeshed proposal:

* If a class declares no constructors, you get the default one implicitly. This optimizes for the data class case.

* If a class does declare a constructor, or you wish to rename the default constructor, you can add it explicitly using `= default;`:

    ```dart
    class Point {
      num x, y;
      Point.cartesian = default;
    }
    ```

    (This might also give you a notation for making it a *const* constructor if desired.)

Reply:
There is not just *one* default constructor if your superclass has multiple generative constructors. You get initializing forwarding constructors for all of them.

Suggested syntax for default constructor:
```dart
default Point.cartesian; // no super, so use `:super()`.
```
or, if you want to forward to specify the super-constructor.
```dart
default ColorPoint: super.cartesian;
```

I guess we could say that:
* If your class has no constructors, you get the full forwarding initializing constructors.
* If your class declares at least one constructor, no explicit default constructor, no unnamed constructor, and your superclass has an unnamed constructor, you get a default forwarding unnamed constructor `default MyName: super;`.

There's still no way to *avoid* that short of declaring `default MyName._preventDefault;`. I'm not happy about that. I think I'd prefer to just make you write `default MyName;` in that case, and not give you a default constructor if you declare *any* constructor explicitly.

Reply:
> I think I'd prefer to just make you write `default MyName;` in that case, and not give you a default constructor if you declare _any_ constructor explicitly.

+1.